### PR TITLE
Improve get_set_cell_value_code

### PR DESCRIPTION
**Pull request by NeeEoo**

### CODE REVIEW

File: Compiler/General.py [hunk 1]
```diff
@@ -110,32 +114,41 @@ def get_set_cell_value_code(new_value, previous_value, zero_next_cell_if_necessa
     def get_abc(offset):
         # returns a,b,c such that a*b+c=offset and a+b+c is minimal
 
-        min_a, min_b, min_c = offset, 1, 0
-        min_sum = min_a + min_b + min_c
+        min_a, min_b, min_c = 0, 0, 0
```
Comment (elikaski) [inline Compiler/General.py:117]: if offset = 0/1/2/3 then we won't run the loop and these will be zeros

File: Compiler/General.py [hunk 2]
```diff
@@ -110,32 +114,41 @@ def get_set_cell_value_code(new_value, previous_value, zero_next_cell_if_necessa
...
+        left = 1
+        right = offset // 2 - 1
 
-        for i in range(1, offset // 2 + 1):
-            a, b, c = i, offset // i, offset % i
-            curr_sum = a + b + c
+        while right >= left:
+            a, b = left+1, right+1
+            c = offset - a * b
+            curr_sum = abs(a) + abs(b) + abs(c)
```
Comment (elikaski) [inline Compiler/General.py:126]: Can a,b be negative?
I think that adding negative values (c) only makes it more complicated :(

If you want to cover the case where there can be <a> * <b> plus signs, followed by <c> minus signs, then the previous way was covering this by decreasing <b> by 1 and keeping only plus signs (just an example)
(I thought about this algorithm a lot)

File: Compiler/General.py [hunk 3]
```diff
@@ -99,8 +99,12 @@ def get_set_cell_value_code(new_value, previous_value, zero_next_cell_if_necessa
     # after the code of this function is executed, the pointer will point to the original cell
     # this function returns the shorter code between "naive" and "looped"
 
+    def get_char(value):
```
Comment (elikaski) [inline Compiler/General.py:102]: I think a function is redundant here

**Review by elikaski — COMMENTED** [review]
I think that leaving the old way of "set cell value" code would be better
The get_print_string_code usage is good though :)

File: Compiler/General.py (hunk 2 above)
Comment (NeeEoo) [reply Compiler/General.py:126]: Only b will be negative if offset is negative. And if offset is negative c is inverted

Comment (NeeEoo) [reply Compiler/General.py:126]: a and b inside of get_abc can never be negative

**Comment by NeeEoo** [comment]
Here is my test code plus the result of it. https://gist.github.com/NeeEoo/1057b888f74f85fe22cfb36de89ffd73

File: Compiler/General.py (hunk 1 above)
Comment (NeeEoo) [reply Compiler/General.py:117]: Will fix this. Even if naive wins and this never really happens.

File: Compiler/General.py (hunk 3 above)
Comment (NeeEoo) [reply Compiler/General.py:102]: The function is used twice

**Comment by NeeEoo** [comment]
Which would you prefer, or do you have a better way to do this

```py
looped += get_char(-c if is_negative else c) * abs(c)  # c more actions
```
```py
looped += get_char(c * (int(is_negative) * -2 + 1)) * abs(c)  # c more actions
```

`(int(is_negative) * -2 + 1)` returns -1 if is_negative is true otherwise return 1

I have fixed the issue where the loop wont run if offset == 0/1/2/3. But i haven't pushed it yet.

**Comment by elikaski** [comment]
OK, following your link to your getset.py, I'm convinced your code is better than mine 😄 
I reviewed it again and now I understand it
It looks OK to me, I'm ready to merge

Regarding your question,
> ```python
> looped += get_char(-c if is_negative else c) * abs(c)  # c more actions
> ```
I think this one is more readable, but choose whatever you prefer

Please let me know when you're ready for me to merge
Thank you :)

**Comment by elikaski** [comment]
Thanks! Your change increases performance and makes the generated code shorter :)

**Pull request merged**